### ISSUE TITLE
Querying comments via relationships does not work

### ISSUE THREAD

**Issue by SirLamer** [cloudcreativity/demo-laravel-json-api#4]
(NOTE: While writing this I solved the problem, but submitted to document it while I prepare a PR.)

I am tinkering with this demo in order to improve my understanding.

When I go to the API URL "http://demo.dev/api/v1/posts/1/comments" (with demo.dev being my test domain) I get a "Internal Server Error" response. The error log is the following:

```
[2017-02-26 06:46:12] local.ERROR: Symfony\Component\Debug\Exception\FatalThrowableError: Type error: Argument 3 passed to Neomerx\JsonApi\Schema\SchemaProvider::createRelationshipObject() must be of the type array, object given, called in /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Schema/SchemaProvider.php on line 220 in /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Schema/SchemaProvider.php:319
Stack trace:
#0 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Schema/SchemaProvider.php(220): Neomerx\JsonApi\Schema\SchemaProvider->createRelationshipObject(Object(App\Comment), 'post', Object(App\Post))
#1 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Encoder/Parser/Parser.php(174): Neomerx\JsonApi\Schema\SchemaProvider->getRelationshipObjectIterator(Object(App\Comment), true, Array)
#2 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Encoder/Parser/Parser.php(132): Neomerx\JsonApi\Encoder\Parser\Parser->parseData()
#3 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Encoder/Encoder.php(219): Neomerx\JsonApi\Encoder\Parser\Parser->parse(Object(Illuminate\Database\Eloquent\Collection))
#4 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Encoder/Encoder.php(156): Neomerx\JsonApi\Encoder\Encoder->encodeDataToArray(Object(Neomerx\JsonApi\Schema\Container), Object(Illuminate\Database\Eloquent\Collection), Object(Neomerx\JsonApi\Encoder\Parameters\EncodingParameters))
#5 /home/vagrant/Code/demo-laravel-json-api/vendor/neomerx/json-api/src/Http/Responses.php(94): Neomerx\JsonApi\Encoder\Encoder->encodeData(Object(Illuminate\Database\Eloquent\Collection), Object(Neomerx\JsonApi\Encoder\Parameters\EncodingParameters))
#6 /home/vagrant/Code/demo-laravel-json-api/vendor/cloudcreativity/json-api/src/Http/Responses/ResponseFactory.php(97): Neomerx\JsonApi\Http\Responses->getContentResponse(Object(Illuminate\Database\Eloquent\Collection), 200, Array, NULL, Array)
#7 /home/vagrant/Code/demo-laravel-json-api/vendor/cloudcreativity/laravel-json-api/src/Http/Controllers/EloquentController.php(187): CloudCreativity\JsonApi\Http\Responses\ResponseFactory->content(Object(Illuminate\Database\Eloquent\Collection))
#8 [internal function]: CloudCreativity\LaravelJsonApi\Http\Controllers\EloquentController->readRelatedResource(Object(CloudCreativity\JsonApi\Http\Requests\Request), '1', 'comments', 'posts')
#9 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(55): call_user_func_array(Array, Array)
#10 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(44): Illuminate\Routing\Controller->callAction('readRelatedReso...', Array)
#11 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Route.php(203): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(App\Http\Controllers\Api\PostsController), 'readRelatedReso...')
#12 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Route.php(160): Illuminate\Routing\Route->runController()
#13 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(559): Illuminate\Routing\Route->run()
#14 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#15 /home/vagrant/Code/demo-laravel-json-api/vendor/cloudcreativity/laravel-json-api/src/Http/Middleware/HandleRequest.php(64): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#16 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(148): CloudCreativity\LaravelJsonApi\Http\Middleware\HandleRequest->handle(Object(Illuminate\Http\Request), Object(Closure), 'App\\JsonApi\\Pos...')
#17 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#18 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Middleware/ThrottleRequests.php(49): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#19 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(148): Illuminate\Routing\Middleware\ThrottleRequests->handle(Object(Illuminate\Http\Request), Object(Closure), '60', '1')
#20 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#21 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(46): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#22 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(148): Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#23 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#24 /home/vagrant/Code/demo-laravel-json-api/vendor/cloudcreativity/laravel-json-api/src/Http/Middleware/BootJsonApi.php(91): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#25 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(148): CloudCreativity\LaravelJsonApi\Http\Middleware\BootJsonApi->handle(Object(Illuminate\Http\Request), Object(Closure), 'v1')
#26 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#27 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#28 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(561): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#29 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(520): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#30 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(498): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#31 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(174): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#32 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#33 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#34 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(149): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#35 /home/vagrant/Code/demo-laravel-json-api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(116): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#36 /home/vagrant/Code/demo-laravel-json-api/public/index.php(53): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#37 {main}  
```

I traced this through and found a couple key points of interest.

See this method 'Neomerx\JsonApi\Schema\SchemaProvider'
```php
    /**
     * @inheritdoc
     */
    public function getRelationshipObjectIterator($resource, $isPrimary, array $includeRelationships)
    {
        foreach ($this->getRelationships($resource, $isPrimary, $includeRelationships) as $name => $desc) {
            yield $this->createRelationshipObject($resource, $name, $desc);
        }
    }
```

The "$desc" in the foreach loop is passing an Eloquent model but the following method is expecting an associative array:
```php
    protected function createRelationshipObject($resource, $name, array $desc)
    {
        $data          = $this->getValue($desc, self::DATA);
        $meta          = $this->getValue($desc, self::META, null);
        $isShowSelf    = ($this->getValue($desc, self::SHOW_SELF, false) === true);
        $isShowRelated = ($this->getValue($desc, self::SHOW_RELATED, false) === true);
        $isShowData    = ($this->getValue($desc, self::SHOW_DATA, array_key_exists(self::DATA, $desc)) === true);
        $links         = $this->readLinks($resource, $name, $desc, $isShowSelf, $isShowRelated);

        return $this->factory->createRelationshipObject($name, $data, $links, $meta, $isShowData, false);
    }
```

Oooh... while writing this I figured out the issue, I'll look at submitting a PR.

Edit to Comments\Schema:
```php
    /**
     * @param object $resource
     * @param bool $isPrimary
     * @param array $includeRelationships
     * @return array
     */
    public function getRelationships($resource, $isPrimary, array $includeRelationships)
    {
        if (!$resource instanceof Comment) {
            throw new RuntimeException('Expecting a comment model.');
        }

        return [
            'post' => [
                self::SHOW_SELF => true,
                self::SHOW_RELATED => true,
                self::DATA => isset($includeRelationships['post']) ?
                    $resource->post : $this->createBelongsToIdentity($resource, 'post'),
            ],
            'created-by' => [
                self::SHOW_SELF => true,
                self::SHOW_RELATED => true,
                self::DATA => isset($includeRelationships['created-by']) ?
                    $resource->person : $this->createBelongsToIdentity($resource, 'person'),
            ],
        ];

        /*
        return [
            'post' => isset($includeRelationships['post']) ?
                $resource->post : $this->createBelongsToIdentity($resource, 'post'),
            'created-by' => isset($includeRelationships['created-by']) ?
                $resource->person : $this->createBelongsToIdentity($resource, 'person'),
        ];
        */
    }
```

**Comment by SirLamer** [comment]
This is corrected on the 5.3 branch in this PR:
https://github.com/cloudcreativity/demo-laravel-json-api/pull/6

If you accept, replicate to 5.4 branch as well, please.

**Comment by lindyhopchris** [comment]
@SirLamer will do! Thanks
